gobblin-metrics/src/main/java/gobblin/metrics/RootMetricContext.java

```diff
@@ -0,0 +1,142 @@
+
+  public static final String ROOT_METRIC_CONTEXT = "RootMetricContext";
+
+  @Getter
+  private final ReferenceQueue<MetricContext> referenceQueue;
+  private final Set<InnerMetricContext> innerMetricContexts;
+  private final ScheduledExecutorService referenceQueueExecutorService;
+
+  private final Set<ContextAwareReporter> reporters;
+
+  private boolean reportingStarted;
+
+  private RootMetricContext(List<Tag<?>> tags) throws NameConflictException {
+    super(ROOT_METRIC_CONTEXT, null, tags, true);
+    this.innerMetricContexts = Sets.newConcurrentHashSet();
+    this.referenceQueue = new ReferenceQueue<MetricContext>();
+    this.referenceQueueExecutorService = MoreExecutors.getExitingScheduledExecutorService(new ScheduledThreadPoolExecutor(1,
+        ExecutorsUtils.newThreadFactory(Optional.of(log), Optional.of("GobblinMetrics-ReferenceQueue"))));
+    this.referenceQueueExecutorService.scheduleWithFixedDelay(new CheckReferenceQueue(), 0, 2, TimeUnit.SECONDS);
+
+    this.reporters = Sets.newConcurrentHashSet();
+    this.reportingStarted = false;
+  }
+
+  private static void initialize() {
+    try {
+      INSTANCE = new RootMetricContext(Lists.<Tag<?>>newArrayList());
+    } catch (NameConflictException nce) {
+      // Should never happen, as there is no parent, so no conflict.
+      throw new RuntimeException("Failed to generate root metric context. This is an error in the code.");
+    }
+  }
+
+  /**
+   * Get the singleton {@link RootMetricContext}.
+   * @return singleton instance of {@link RootMetricContext}.
+   */
+  public synchronized static RootMetricContext get() {
+
+    if(INSTANCE == null) {
+      initialize();
+    }
+    return INSTANCE;
+  }
+
+  private static RootMetricContext INSTANCE;
+
+  /**
+   * Checks the {@link ReferenceQueue} to find any {@link MetricContext}s that have been garbage collected, and
+   * sends a {@link MetricContextCleanupNotification} to all targets.
+   */
+  private class CheckReferenceQueue implements Runnable {
+    @Override public void run() {
+      Reference<? extends MetricContext> reference;
+      while((reference = referenceQueue.poll()) != null) {
+        ContextWeakReference contextReference = (ContextWeakReference)reference;
+
+        sendNotification(new MetricContextCleanupNotification(contextReference.getInnerContext()));
+        innerMetricContexts.remove(contextReference.getInnerContext());
+      }
+    }
+  }
+
+  public void addNewReporter(ContextAwareReporter reporter) {
+    this.reporters.add(this.closer.register(reporter));
+    if(this.reportingStarted) {
+      reporter.start();
+    }
+  }
+
+  public void removeReporter(ContextAwareReporter reporter) {
+    if(this.reporters.contains(reporter)) {
+      reporter.stop();
+      this.reporters.remove(reporter);
+    }
+  }
+
+  public void startReporting() {
+    this.reportingStarted = true;
+    for(ContextAwareReporter reporter : this.reporters) {
+      reporter.start();
```

Comment:
I suggest you catch `Throwable` around `reporter.start()` so the failure to start of one reporter won't cause other reports to not start.